capsules/src/virtual_aes_ccm.rs

```diff
@@ -0,0 +1,802 @@
+enum CCMState {
+    Idle,
+    Auth,
+    Encrypt,
+}
+
+// to cache up the function parameters of the crypt() function
+struct CryptFunctionParameters {
+    buf: &'static mut [u8],
+    a_off: usize,
+    m_off: usize,
+    m_len: usize,
+    mic_len: usize,
+    confidential: bool,
+    encrypting: bool,
+}
+
+impl CryptFunctionParameters {
+    pub fn new(
+        buf: &'static mut [u8],
+        a_off: usize,
+        m_off: usize,
+        m_len: usize,
+        mic_len: usize,
+        confidential: bool,
+        encrypting: bool,
+    ) -> CryptFunctionParameters {
+        CryptFunctionParameters {
+            buf: buf,
+            a_off: a_off,
+            m_off: m_off,
+            m_len: m_len,
+            mic_len: mic_len,
+            confidential: confidential,
+            encrypting: encrypting,
+        }
+    }
+}
+
+pub struct MuxAES128CCM<'a, A: AES128<'a> + AES128Ctr + AES128CBC> {
+    aes: &'a A,
+    clients: List<'a, VirtualAES128CCM<'a, A>>,
+    inflight: OptionalCell<&'a VirtualAES128CCM<'a, A>>,
+    // debug_counter: Cell<i32>,
+}
+
+impl<'a, A: AES128<'a> + AES128Ctr + AES128CBC> MuxAES128CCM<'a, A> {
+    pub fn new(aes: &'a A) -> MuxAES128CCM<'a, A> {
+        // aes.enable(); // enable the hardware
+        // debug!("new mux");
+        MuxAES128CCM {
+            aes: aes,
+            clients: List::new(),
+            inflight: OptionalCell::empty(),
+            // debug_counter: Cell::new(0),
+        }
+    }
+
+    /// manually re-enable the hardware
+    pub fn enable(&self) {
+        self.aes.enable();
+    }
+
+    /// disable the underlying hardware
+    pub fn disable(&self) {
+        self.aes.disable();
+    }
+
+    fn do_next_op(&self) {
+        if self.inflight.is_none() {
+            let mnode = self.clients.iter().find(|node| node.queued_up.is_some());
+            mnode.map(|node| {
+                self.inflight.set(node);
+                let parameters: CryptFunctionParameters = node.queued_up.take().unwrap();
+                let (res, _) = node.crypt_r(parameters); // eats the parameters
+
+                // notice that we didn't put the parameters back...
+                // because it's already eaten
+                if res != ReturnCode::SUCCESS {
+                    if node.crypt_client.is_none() {
+                        debug!(
+                            "virtual_aes_ccm: no crypt_client is registered in VirtualAES128CCM"
+                        );
+                    }
+                    if node.buf.is_none() {
+                        debug!("virtual_aes_ccm: no buffer is binded with VirtualAES128CCM");
+                    }
+                    // notify the client that there's a failure
+                    node.buf.take().map(|buf| {
+                        node.crypt_client.map(move |client| {
+                            client.crypt_done(buf, res, false);
+                        });
+                    });
```

Comment:
Unfortunately it is problematic to have an error callback here. The issue is that when the user calls `crypt()` and nothing else is using AES, do_next_op runs immediately. If there is an error, this callback is called synchronously. However, tock requires that callbacks _only_ be generated stemming from interrupts.

The general way to solve this is to use a deferred call in `crypt()` so that do_next_op is only ever called from a callback context. The virtual_uart driver illustrates this.

Reply:
Thanks Brad, I've now modified it to use dynamic deferred call to call `do_next_op()` in `crypt()`

Reply:
> I think this should either be merged with the existing aes_ccm capsule, or replace it. The virtualizer layer shouldn't have to duplicate all of the logic from the aes_ccm capsule.

Yeah, there should be further transitions to use the virtualized driver. But currently I couldn't perform such transition as aes_ccm.rs is used somewhere else. Probably we should do the transition later. Anyways, the virtual_aes_ccm is also functional.